Enhance docs around Gateways, Routes and TLS (#633)
kubernetes-sigs/gateway-api: site-src/guides/tls.md

With Gateway API, TLS configuration of downstream and
upstream connections is managed independently.

Depending on the Listener Protocol, different TLS modes and Route types are supported.

Listener Protocol | TLS Mode | Route Type Supported
--- | --- | ---

> TLS | Passthrough | TLSRoute

Reviewer: Should it also be possible to have protocol `TLS`, tls mode `Terminate`, with route type `TLSRoute`?

Author: Given the Slack discussion, I understood it as "No". As TLSRoute will need TLS routing information which would be terminated? But maybe im wrong.

Reviewer: Nah sounds like I don't understand either. I think the bit I'm not grasping is what @howardjohn described in the thread in regards to the 3 protocols, `incoming protocol, routing protocol, and upstream protocol` and how those apply.

Author: Good questions. Maybe @howardjohn @robscott @youngnick can help?

Reviewer 2: I think there are three protocols involved here:

Protocol to the Gateway - defined by the Listener
Protocol to be used for Routing - defined by the Route (selected by the Listener)
Protocol to be used upstream from the Gateway - defined by the upstream Service, usually the same as routing protocol

That's at least my interpretation of how the different pieces fit together. We likely need a follow up to better document these concepts.

Reviewer 3: > I wonder then also how the AppProtocol field comes into play, meaning if you had a TCPProxy with TLS on the gateway and an AppProtocol of TLS would you terminate TLS at the gateway, then forward upstream TLS?

I would expect two valid configurations for AppProtocol=TLS personally:
* TLS=Passthrough, forward the original TLS
* TLS=Terminate, with BackendPolicy configuring to use TLS. This would terminate then originate a new TLS connection.

Author: So should I change and list both modes?

Reviewer 2: I could be wrong, but I think this discussion is somewhat different than the kind of routes that are supported with each mode. So although I think this is important to document, I don't think we need to delay getting this PR in.

Author: SGTM. Nitpick changes coming in soon

[TLS Certificate in Route](#tls-certificate-in-route) provides an example
of how this feature can be used.

> Also, as mentioned above, the Route Kind (`HTTPRoute`, `TLSRoute`, `TCPRoute`) is dependant on the protocol on the listener level. Listener Protocol `HTTPS` or `HTTP` allows to bind against `HTTPRoute` as the TLS Termination is done at the listener level and thus, only http information are used for routing.

Reviewer 2: Nit: A couple small typos
```suggestion
Also, as mentioned above, the Route Kind (`HTTPRoute`, `TLSRoute`, `TCPRoute`) is dependent on the protocol on the listener level. Listeners with `HTTPS` or `HTTP` protocols can use `HTTPRoute` as the TLS Termination is done at the listener level and thus, only HTTP information is used for routing.
```

> Also, as mentioned above, the Route Kind (`HTTPRoute`, `TLSRoute`, `TCPRoute`) is dependant on the protocol on the listener level. Listener Protocol `HTTPS` or `HTTP` allows to bind against `HTTPRoute` as the TLS Termination is done at the listener level and thus, only http information are used for routing.

Reviewer 2: Also one more tiny nit - do you mind wrapping these lines at 80 characters?

Author: Done

> For `TLSRoute`, its allowed to bind against `TCP` and `TLS` depending on the mode. If the mode is `Terminate`, plain `TCP` routing is possible. If its `TLS`, its also possible to forward all TLS information for `TLS` Routing.

Reviewer 2: ```suggestion
Listeners with the `TLS` protocol must use `TLSRoute` when the mode is set to `Passthrough` and `TCPRoute` when the mode is `Terminate`.
```

> For `TCPRoute`, only plain TCP routing (e.g. no SNI) is possible.

Reviewer 2: Listeners with the `TCP` protocol must use `TCPRoute`.
